### PR TITLE
add getspent RPCs for bitcore block explorer

### PR DESCRIPTION
**Pull request by LarryRuane**
Part of #3708.

### CODE REVIEW

**Comment by LarryRuane** [comment]
Please look only at the commits from April 15 and beyond (the earlier ones are part of other PRs).

**Comment by mms710** [comment]
Note for reviewers: these PRs must be reviewed and merged in this order:
#3863 
#3926 
#3953 
#3972 

Between every merge, ping Larry to rebase the next PR in the list so you can review and merge the next one down the line

File: src/rpc/misc.cpp [hunk 1]
```diff
@@ -924,6 +929,51 @@ UniValue getaddresstxids(const UniValue& params, bool fHelp)
...
+            "\nExamples:\n"
+            + HelpExampleCli("getspentinfo", "'{\"txid\": \"0437cd7f8525ceed2324359c2d0ba26006d92d856a9c20fa0241106ee5a597c9\", \"index\": 0}'")
+            + HelpExampleRpc("getspentinfo", "{\"txid\": \"0437cd7f8525ceed2324359c2d0ba26006d92d856a9c20fa0241106ee5a597c9\", \"index\": 0}")
+        );
+
+    UniValue txidValue = find_value(params[0].get_obj(), "txid");
+    UniValue indexValue = find_value(params[0].get_obj(), "index");
+
+    if (!txidValue.isStr() || !indexValue.isNum()) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid txid or index");
```
Comment (Eirik0) [inline src/rpc/misc.cpp]: This may as well be two separate if statements/error messages.

File: src/rpc/misc.cpp [hunk 2]
```diff
@@ -924,6 +929,51 @@ UniValue getaddresstxids(const UniValue& params, bool fHelp)
 
 }
 
+// insightexplorer
+UniValue getspentinfo(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() != 1 || !params[0].isObject())
+        throw runtime_error(
+            "getspentinfo\n"
```
Comment (Eirik0) [inline src/rpc/misc.cpp]: I suppose this documentation should be updated in a similar fashion to #3926

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -313,6 +402,30 @@ UniValue getrawmempool(const UniValue& params, bool fHelp)
     return mempoolToJSON(fVerbose);
 }
 
+// insightexplorer
+UniValue getblockdeltas(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() != 1)
+        throw runtime_error("");
```
Comment (Eirik0) [inline src/rpc/blockchain.cpp]: Is this the documentation for this RPC? >.<

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -125,6 +126,94 @@ UniValue blockheaderToJSON(const CBlockIndex* blockindex)
...
+            const CTxOut &out = tx.vout[k];
+            UniValue delta(UniValue::VOBJ);
+            const uint160 addrhash = out.scriptPubKey.AddressHash();
+            CTxDestination dest;
+
+            if (out.scriptPubKey.IsPayToScriptHash()) {
+                dest = CScriptID(addrhash);
+            } else if (out.scriptPubKey.IsPayToPublicKeyHash()) {
+                dest = CKeyID(addrhash);
+            }
```
Comment (Eirik0) [inline src/rpc/blockchain.cpp]: What happens if `dest` is not set?

File: qa/rpc-tests/spentindex.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        self.nodes[0].generate(105)
+        self.sync_all()
+
+        chain_height = self.nodes[1].getblockcount()
+        assert_equal(chain_height, 105)
+
+        # Test getrawtransaction changes and the getspentinfo RPC
+
+        # send coinbase to address a
+        a = self.nodes[1].getnewaddress()
```
Comment (Eirik0) [inline qa/rpc-tests/spentindex.py]: Nit: I prefer `addr1`/`addr2` to `a`/`b`

Comment (Eirik0) [reply qa/rpc-tests/spentindex.py]: I guess then `_a` and `_b` change to `_1` and `_2`.
_shrug_

File: qa/rpc-tests/spentindex.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        assert('spentTxId' not in tx_b['vout'][0])
+        assert('spentIndex' not in tx_b['vout'][0])
+        assert('spentHeight' not in tx_b['vout'][0])
+        assert_equal(tx_b['height'], 107)
+
+        # Given a transaction output, getspentinfo() returns a reference
+        # to the (later, confirmed) transaction that spent that output,
+        # that is, the transaction that used this output as an input.
+        spentinfo = self.nodes[2].getspentinfo({
+            'txid': txid_a, 'index': n})
```
Comment (Eirik0) [inline qa/rpc-tests/spentindex.py]: Nit: We don't really need to line wrap here.

File: qa/rpc-tests/spentindex.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        assert_equal(spentinfo['txid'], txid_b)
+
+        # Some invalid argument checks...
+        # must specify a valid txid and a valid index
+        error = ""
+        try:
+            self.nodes[1].getspentinfo({'txid': txid_a, 'index': 2})
+        except JSONRPCException, e:
+            error = e.error['message']
+        assert_equal(error, "Unable to get spent info")
```
Comment (Eirik0) [inline qa/rpc-tests/spentindex.py]:
```
try:
    self.nodes[1].getspentinfo({'txid': txid_a, 'index': 2})
    fail('getspentinfo should have thrown an exception')
except JSONRPCException, e:
    assert_equal(e.error['message'], "Unable to get spent info")
```

File: qa/rpc-tests/spentindex.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+        assert_equal(len(to_b_tx['inputs']), 1)
+        assert_equal(to_b_tx['inputs'][0]['index'], 0)
+        assert_equal(to_b_tx['inputs'][0]['prevtxid'], txid_a)
+        assert_equal(to_b_tx['inputs'][0]['satoshis'], -2*COIN)
+
+        assert_equal(len(to_b_tx['outputs']), 2)
+        # find the nonchange output, which is the payment to b
+        out = filter(lambda o: o['satoshis'] == 1*COIN, to_b_tx['outputs'])
+        assert_equal(out[0]['address'], b)
```
Comment (Eirik0) [inline qa/rpc-tests/spentindex.py]: We could assert that the size of out is correct. Same above.

File: src/rpc/blockchain.cpp (hunk 3 above)
Comment (Eirik0) [reply src/rpc/blockchain.cpp]: I see this is added in #3972.

Comment (LarryRuane) [reply src/rpc/blockchain.cpp]: I wish I'd noticed your comment before I implemented this -- again! Oh well, I think I did it better this time.

File: src/rpc/blockchain.cpp (hunk 4 above)
Comment (LarryRuane) [reply src/rpc/blockchain.cpp]: It's okay, because `IsValidDestination(dest)` will return `false`.

**Comment by LarryRuane** [comment]
Thanks, @Eirik0, commit https://github.com/zcash/zcash/pull/3953/commits/74ecb8c232df82c20ed45f3b7ed8a3012a37ba71 should address all of your comments.

**Comment by LarryRuane** [comment]
> I had ACKed, but then I was thinking that the RPCs added should probably include the same experimental feature checks as suggested by @str4d: [#3926 (comment)](https://github.com/zcash/zcash/pull/3926#discussion_r305898279)

@Eirik0, thanks, good point, plus, I should probably add unit tests (for argument handling, including what you mentioned), similar to https://github.com/zcash/zcash/pull/3926/commits/38bcb0e8e74d6f7e6ad625ece2e3ecacb62bdc0a#diff-b89234789a9a9ee29cb0758f69a73133R363

**Comment by LarryRuane** [comment]
@Eirik0, @str4d, I just rebased and added a commit to improve the testing. AFAIK this is ready to be merged except for squashing.

File: qa/rpc-tests/spentindex.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+        # Given a transaction output, getspentinfo() returns a reference
+        # to the (later, confirmed) transaction that spent that output,
+        # that is, the transaction that used this output as an input.
+        spentinfo = self.nodes[2].getspentinfo({'txid': txid1, 'index': n})
+        assert_equal(spentinfo['height'], 107)
+        assert_equal(spentinfo['index'], 0)
+        assert_equal(spentinfo['txid'], txid2)
+
+        # Some invalid argument checks...
```
Comment (LarryRuane) [inline qa/rpc-tests/spentindex.py]: I don't know if I need all these checks here ... I wrote them before I was aware of `src/test/rpc_tests.cpp` (you can run the tests I added there as follows: `src/test/test_bitcoin --run_test=rpc_tests/rpc_insightexplorer`). Are these okay, or should I remove them?

File: src/test/rpc_tests.cpp [hunk 10]
```diff
@@ -416,6 +416,14 @@ BOOST_AUTO_TEST_CASE(rpc_insightexplorer)
     CheckRPCThrows("getaddresstxids {\"addresses\":[],\"start\":2,\"end\":3,\"chainInfo\":true}",
         "Start or end is outside chain range");
```
Comment (LarryRuane) [inline src/test/rpc_tests.cpp]: (As I said elsewhere in this review), you can run these tests as follows:
```
$ src/test/test_bitcoin --run_test=rpc_tests/rpc_insightexplorer
```

File: src/main.cpp [hunk 11]
```diff
@@ -1621,6 +1621,8 @@ bool GetSpentIndex(CSpentIndexKey &key, CSpentIndexValue &value)
     AssertLockHeld(cs_main);
     if (!fSpentIndex)
         return false;
+    if (mempool.getSpentIndex(key, value))
```
Comment (LarryRuane) [inline src/main.cpp]: Missed this from upstream: https://github.com/zcash-hackworks/zcash-patched-for-explorer/blob/dffb2ebbca44e475171575092db5c54349b3d10c/src/main.cpp#L1634. @garethtdavies noticed this bug and mentioned it here: https://github.com/zcash/zcash/pull/3972#issuecomment-518052328 (thanks, Gareth!) but I thought it's better to fix it in the earliest possible PR.

File: qa/rpc-tests/spentindex.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
+#!/usr/bin/env python2
+# Copyright (c) 2014-2015 The Bitcoin Core developers
```
Comment (str4d) [inline qa/rpc-tests/spentindex.py]: Was this test backported, or written by us? If the latter, update this header.

File: qa/rpc-tests/spentindex.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+from test_framework.util import start_nodes, stop_nodes, connect_nodes
+from test_framework.util import wait_bitcoinds
+
+from test_framework.mininode import COIN
+
+class SpentIndexTest(BitcoinTestFramework):
+
+    def setup_chain(self):
+        print("Initializing test directory "+self.options.tmpdir)
+        initialize_chain_clean(self.options.tmpdir, 4)
```
Comment (str4d) [inline qa/rpc-tests/spentindex.py]:
```suggestion
        initialize_chain_clean(self.options.tmpdir, 3)
```

File: src/rpc/rawtransaction.cpp [hunk 14]
```diff
@@ -197,6 +197,7 @@ void TxToJSON(const CTransaction& tx, const uint256 hashBlock, UniValue& entry)
         UniValue out(UniValue::VOBJ);
         out.push_back(Pair("value", ValueFromAmount(txout.nValue)));
         out.push_back(Pair("valueZat", txout.nValue));
+        out.push_back(Pair("valueSat", txout.nValue));
```
[the remark this answers is not in the document]
Comment (LarryRuane) [reply src/rpc/rawtransaction.cpp]: I know it's not ideal, but, yes, the bitcore insight explorer refers to this name, and I wanted this PR (actually, all of #3708) to make our main repo zcashd a drop-in replacement for the hackworks patched version. This seems like something we can do later (reduce tech debt kind of thing).

File: src/rpc/blockchain.cpp [hunk 15]
```diff
@@ -125,6 +126,94 @@ UniValue blockheaderToJSON(const CBlockIndex* blockindex)
...
+                CSpentIndexKey spentKey(input.prevout.hash, input.prevout.n);
+
+                if (!GetSpentIndex(spentKey, spentInfo)) {
+                    throw JSONRPCError(RPC_INTERNAL_ERROR, "Spent information not available");
+                }
+                CTxDestination dest = DestFromAddressHash(spentInfo.addressType, spentInfo.addressHash);
+                if (IsValidDestination(dest)) {
+                    delta.push_back(Pair("address", EncodeDestination(dest)));
+                }
+                delta.push_back(Pair("satoshis", -1 * spentInfo.satoshis));
```
[the remark this answers is not in the document]
Comment (LarryRuane) [reply src/rpc/blockchain.cpp]: Yes, that comes from upstream: https://github.com/zcash-hackworks/zcash-patched-for-explorer/blob/v2.0.4-insight-explorer/src/rpc/blockchain.cpp#L178

**Comment by LarryRuane** [comment]
Thanks, @str4d and @Eirik0, I took all of your suggestions (except I'm keeping those duplicate *Zat and *Sat fields), squashed to one commit, ready for r+.

**Comment by Eirik0** [comment]
@zkbot r+

File: src/rpc/rawtransaction.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (str4d) [reply src/rpc/rawtransaction.cpp]: Given that `-insightexplorer` is being added as an experimental feature, we have the ability to do this later, so I'm fine with leaving this as-is for now.

**Comment by str4d** [comment]
The failures are material (an insightexplorer-related RPC test that fails, and the `spentindex` RPC test added in this PR that passes but doesn't import `fail`, so it would crash in failure cases).

File: qa/rpc-tests/addressindex.py [hunk 16]
```diff
@@ -202,6 +202,11 @@ def check_balance(node_index, address, expected_balance, expected_received=None)
         # a single address can be specified as a string (not json object)
         assert_equal([mempool[1]], self.nodes[0].getaddressmempool(addr1))
 
+        tx = self.nodes[0].getrawtransaction(txid, 1)
+        assert_equal(tx['vin'][0]['address'], addr1)
+        assert_equal(tx['vin'][0]['value'], 4)
+        assert_equal(tx['vin'][0]['valueSat'], 4 * COIN)
```
[the remark this answers is not in the document]
Comment (daira) [reply qa/rpc-tests/addressindex.py]: Alternative suggestion: also add
```
        assert_equal(tx['vin'][0]['valueZat'], 4 * COIN)
```

File: src/rpc/blockchain.cpp [hunk 17]
```diff
@@ -125,6 +126,94 @@ UniValue blockheaderToJSON(const CBlockIndex* blockindex)
...
+
+            if (out.scriptPubKey.IsPayToScriptHash()) {
+                dest = CScriptID(addrhash);
+            } else if (out.scriptPubKey.IsPayToPublicKeyHash()) {
+                dest = CKeyID(addrhash);
+            }
+            if (IsValidDestination(dest)) {
+                delta.push_back(Pair("address", EncodeDestination(dest)));
+            }
+            delta.push_back(Pair("satoshis", out.nValue));
```
Comment (daira) [inline src/rpc/blockchain.cpp]: Maybe this should be duplicated as "valueZat"? Does not block.

**Comment by LarryRuane** [comment]
Thanks, @str4d, @daira, and @Eirik0, the CI should pass now, so someone could do an r+. (I squashed the test fixes since they were so trivial.)

**Comment by daira** [comment]
@zkbot r+

**Comment by daira** [comment]
Just for future reference, it's sometimes better not to squash changes even when they are trivial, in order to make it easier to review the last changes. Basically just act to minimize the amount of work needed for review, since that's the bottleneck for engineering resources.

**Comment by Eirik0** [comment]
@zkbot retry

**Comment by Eirik0** [comment]
@zkbot retry

**Comment by str4d** [comment]
@zkbot retry

**Pull request merged**